File: textiles_and_garments/textiles_and_garments/report/employee_checkin/employee_checkin.py
# Copyright (c) 2026, Aravind and contributors
# For license information, please see license.txt

# employee_first_checkin.py

import frappe
from frappe import _
from frappe.utils import getdate, today, get_datetime

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    """Define report columns"""
    return [
        {
            "fieldname": "employee",
            "label": _("Employee ID"),
            "fieldtype": "Link",
            "options": "Employee",
            "width": 120
        },
        {
            "fieldname": "employee_name",
            "label": _("Employee Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "department",
            "label": _("Department"),
            "fieldtype": "Link",
            "options": "Department",
            "width": 180
        },
        {
            "fieldname": "designation",
            "label": _("Designation"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "checkin_time",
            "label": _("Check-in Time"),
            "fieldtype": "Datetime",
            "width": 160
        },
        {
            "fieldname": "shift",
            "label": _("Shift"),
            "fieldtype": "Link",
            "options": "Shift Type",
            "width": 180
        },
        {
            "fieldname": "shift_start",
            "label": _("Shift Start Time"),
            "fieldtype": "Datetime",
            "width": 160
        },
        {
            "fieldname": "shift_end",
            "label": _("Shift End Time"),
            "fieldtype": "Datetime",
            "width": 160
        },
        {
            "fieldname": "time_difference",
            "label": _("Early/Late (Minutes)"),
            "fieldtype": "Int",
            "width": 130
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 100
        }
    ]

def get_data(filters):
    """
    Fetch first check-in for each employee on a daily basis
    """
    conditions = get_conditions(filters)
    
    # Get first check-in per employee per day
    data = frappe.db.sql(f"""
        SELECT 
            ec.employee,
            ec.employee_name,
            ec.department,
            ec.designation,
            ec.checkin_time,
            ec.shift,
            ec.shift_start,
            ec.shift_end,
            ec.time_difference,
            ec.status
        FROM (
            SELECT 
                checkin.employee as employee,
                checkin.employee_name as employee_name,
                emp.department as department,
                emp.designation as designation,
                MIN(checkin.time) as checkin_time,
                checkin.shift as shift,
                checkin.shift_start as shift_start,
                checkin.shift_end as shift_end,
                TIMESTAMPDIFF(MINUTE, checkin.shift_start, MIN(checkin.time)) as time_difference,
                CASE 
                    WHEN TIMESTAMPDIFF(MINUTE, checkin.shift_start, MIN(checkin.time)) <= 0 THEN 'On Time'
                    WHEN TIMESTAMPDIFF(MINUTE, checkin.shift_start, MIN(checkin.time)) <= 15 THEN 'Grace Period'
                    ELSE 'Late'
                END as status,
                DATE(checkin.time) as checkin_date
            FROM 
                `tabEmployee Checkin` checkin
            LEFT JOIN 
                `tabEmployee` emp ON emp.name = checkin.employee
            WHERE 
                1=1
                {conditions}
            GROUP BY 
                checkin.employee, DATE(checkin.time), checkin.shift, checkin.shift_start, 
                checkin.shift_end, emp.department, emp.designation, checkin.employee_name
        ) ec
        ORDER BY 
            ec.checkin_date DESC, ec.checkin_time ASC
    """, filters, as_dict=1)
    
    return data

def get_conditions(filters):
    """Build additional filter conditions"""
    conditions = ""
    
    if filters.get("from_date"):
        conditions += " AND DATE(checkin.time) >= %(from_date)s"
    else:
        # Default to today if no date filter
        conditions += f" AND DATE(checkin.time) = '{today()}'"
    
    if filters.get("to_date"):
        conditions += " AND DATE(checkin.time) <= %(to_date)s"
    
    if filters.get("employee"):
        conditions += " AND checkin.employee = %(employee)s"
    
    if filters.get("department"):
        conditions += " AND emp.department = %(department)s"
    
    if filters.get("shift"):
        conditions += " AND checkin.shift = %(shift)s"
    
    if filters.get("company"):
        conditions += " AND emp.company = %(company)s"
    
    return conditions

File: textiles_and_garments/textiles_and_garments/report/employee_checkin/employee_checkin.js
// Copyright (c) 2026, Aravind and contributors
// For license information, please see license.txt



// employee_first_checkin.js

frappe.query_reports["Employee Checkin"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 0
        },
        {
            "fieldname": "department",
            "label": __("Department"),
            "fieldtype": "Link",
            "options": "Department",
            "reqd": 0
        },
        {
            "fieldname": "employee",
            "label": __("Employee"),
            "fieldtype": "Link",
            "options": "Employee",
            "reqd": 0
        },
        {
            "fieldname": "shift",
            "label": __("Shift"),
            "fieldtype": "Link",
            "options": "Shift Type",
            "reqd": 0
        }
    ],
    
    "onload": function(report) {
        // Add Summary button
        report.page.add_inner_button(__("Summary"), function() {
            let data = report.data;
            
            if (!data || data.length === 0) {
                frappe.msgprint(__("No data available to summarize"));
                return;
            }
            
            // Calculate summary statistics
            let total_employees = data.length;
            let on_time = data.filter(row => row.status === 'On Time').length;
            let grace_period = data.filter(row => row.status === 'Grace Period').length;
            let late = data.filter(row => row.status === 'Late').length;
            
            let unique_departments = [...new Set(data.map(row => row.department).filter(Boolean))].length;
            let unique_shifts = [...new Set(data.map(row => row.shift).filter(Boolean))].length;
            
            // Calculate average late time for late employees
            let late_employees = data.filter(row => row.time_difference > 0);
            let avg_late_time = late_employees.length > 0 
                ? late_employees.reduce((sum, row) => sum + (parseFloat(row.time_difference) || 0), 0) / late_employees.length 
                : 0;
            
            // Calculate percentage
            let on_time_percent = ((on_time / total_employees) * 100).toFixed(1);
            let late_percent = ((late / total_employees) * 100).toFixed(1);
            
            // Show summary dialog
            frappe.msgprint({
                title: __("Employee First Check-in - Summary"),
                message: `
                    <div style="font-size: 14px; line-height: 2;">
                        <h4 style="margin-bottom: 15px; color: #2563eb;">📊 Attendance Overview</h4>
                        <table style="width: 100%; border-collapse: collapse;">
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Total Employees:</strong></td>
                                <td style="text-align: right; padding: 8px 0;">${total_employees}</td>
                            </tr>
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Unique Departments:</strong></td>
                                <td style="text-align: right; padding: 8px 0;">${unique_departments}</td>
                            </tr>
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Unique Shifts:</strong></td>
                                <td style="text-align: right; padding: 8px 0;">${unique_shifts}</td>
                            </tr>
                        </table>
                        
                        <h4 style="margin: 20px 0 15px 0; color: #16a34a;">⏰ Punctuality Summary</h4>
                        <table style="width: 100%; border-collapse: collapse;">
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>On Time:</strong></td>
                                <td style="text-align: right; padding: 8px 0; color: #16a34a; font-weight: 700;">${on_time} (${on_time_percent}%)</td>
                            </tr>
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Grace Period:</strong></td>
                                <td style="text-align: right; padding: 8px 0; color: #ea580c;">${grace_period}</td>
                            </tr>
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Late:</strong></td>
                                <td style="text-align: right; padding: 8px 0; color: #dc2626; font-weight: 700;">${late} (${late_percent}%)</td>
                            </tr>
                        </table>
                        
                        <h4 style="margin: 20px 0 15px 0; color: #ea580c;">📈 Performance Metrics</h4>
                        <table style="width: 100%; border-collapse: collapse;">
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Average Late Time:</strong></td>
                                <td style="text-align: right; padding: 8px 0;">${avg_late_time.toFixed(1)} Minutes</td>
                            </tr>
                            <tr style="border-bottom: 1px solid #e5e7eb;">
                                <td style="padding: 8px 0;"><strong>Punctuality Rate:</strong></td>
                                <td style="text-align: right; padding: 8px 0; color: ${on_time_percent >= 90 ? '#16a34a' : '#dc2626'}; font-weight: 700;">${on_time_percent}%</td>
                            </tr>
                        </table>
                    </div>
                `,
                indicator: "blue",
                wide: true
            });
        });
        
        // Add color indicators
        report.page.add_inner_button(__("Refresh"), function() {
            report.refresh();
        });
    },
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Color code the status column
        if (column.fieldname === "status") {
            if (data.status === "On Time") {
                value = `<span style="color: #16a34a; font-weight: 600;">${data.status}</span>`;
            } else if (data.status === "Grace Period") {
                value = `<span style="color: #ea580c; font-weight: 600;">${data.status}</span>`;
            } else if (data.status === "Late") {
                value = `<span style="color: #dc2626; font-weight: 600;">${data.status}</span>`;
            }
        }
        
        // Color code time difference
        if (column.fieldname === "time_difference" && data.time_difference) {
            if (data.time_difference <= 0) {
                value = `<span style="color: #16a34a;">${data.time_difference}</span>`;
            } else if (data.time_difference <= 15) {
                value = `<span style="color: #ea580c;">${data.time_difference}</span>`;
            } else {
                value = `<span style="color: #dc2626; font-weight: 600;">${data.time_difference}</span>`;
            }
        }
        
        return value;
    }
};
